### ISSUE TITLE
Request to add support for NEC PC-98 standard

### ISSUE THREAD

**Issue by Steveneska** [86Box/86Box#3525]
**Is your feature request related to a problem? Please describe.**
This request is not based on an existing problem.

**Describe the solution you'd like**
Since NEC PC-98 is a non-IBM compatible x86 computer standard, many components used by the NEC series of PCs are already emulated in 86Box. Adding support for this series of computer will bring the only cycle accurate option available in an emulator, since other available options like Neko Project-based emulators and DOSBox-X are not. 

**Describe alternatives you've considered**
I've been using Neko Project and DOSBox-X to emulate PC-98 hardware in the past, but emulation accuracy was left to be desired as these emulators could require some workarounds to get software booting. Using 86Box on the other hand got me thinking what a cycle accurate PC-98 emulation would be like.

**Additional context**
If interested, diving into the Neko Project 21/W code could be a good point to start. I can also compile PC-98 research materials if desired.

**Comment by OBattler** [comment]
That's already in our long-term plans.

**Comment by Steveneska** [comment]
Awesome and happy to hear that. Keep up with the great work, I'm looking forward to what's to come.

**Comment by Torinde** [comment]
FM Towns is another non-IBM compatible x86 computer standard. Are there others?

**Comment by cartifanwlr** [comment]
What interests me are older incompatible x86 computers, like Zenith Z-100, RM Nimbus PC-186, Apricot PC and Xen, DEC Rainbow, HP-150 (the touchscreen PC!), etc. Getting these and others working on 86Box would take a lot of work but I think it would be great for hardware preservation.

**Comment by TC1995** [comment]
The DEC Rainbow would also require Z80 emulation in addition to the 8088.
For the NEC PC-98x1, I need help for the timer of the uPD7220 so I can start properly as well as the dip switches.

**Comment by Torinde** [comment]
@cartifanwlr, what I gather so far:
- 8080 and x86: NEC V20/V30 in PC-98, etc.
- Z80 and x86: [NEC μPD9002](https://en.wikipedia.org/wiki/Zilog_Z80#uPD9002) was a hybrid CPU compatible with both Z80 and x86 families - mentioned in [i8080.h](https://github.com/86Box/86Box/blob/5397388a1b1d7376ebe0d4d7e248a5cf95548cfb/src/include/86box/i8080.h#L20) and used in PC-88. Other computers having x86 + Z80 [as two CPUs](https://en.wikipedia.org/wiki/Multiuser_DOS#CCP/M-86/80): [Rainbow 100](https://en.wikipedia.org/wiki/Rainbow_100), [Compupro System 816](https://sr.wikipedia.org/wiki/Compupro_System_816), [LSI Octopus](http://oldcomputer.info/8bit/hilger/index.htm), [LSI M4](https://forum.vcfed.org/index.php?threads/information-wanted-lsi-computers-m-four.1239479/), [CAL PC](https://nosher.net/archives/computers/pcn_1983-05-20_001), most supporting some versions CP/M and DOS
- 8085 and x86 - there were such computers having two CPUs with support for CP/M and DOS: [Zenith Z-100](https://en.wikipedia.org/wiki/Zenith_Z-100), [Compupro System 816](https://sr.wikipedia.org/wiki/Compupro_System_816)

Related:
- FPUs [8231/8232](https://en.wikipedia.org/wiki/Intel_8231/8232) for 8080/8085/Z80. These FPUs were mentioned as suitable for 8086 in its initial documentation
- NEC x87 FPUs discussed at #3399

@TC1995, DOSbox-X emulates PC-98, but I don't know if you can get the uPD7220 details you need from there.

IBM Music Feature Card that includes a Z80 CPU is [emulated as well](https://github.com/joncampbell123/dosbox-x/issues/3596). Code has a comment:
> The reverse-engineered assembly (see contrib/ibm_music_feature_asm/) shows
> there is a way to upload Z80 programs to the IMF card and execute it! Who
> knew that IBM added this back-door :)

Maybe that could be used to run Z80/8080 code when using V20.

**Comment by Steveneska** [comment]
Also related could be the [UNDOCUMENTED 9801/9821 Vol. 2 - Memory and I/O Ports](http://www.webtech.co.jp/company/doc/undocumented_mem/) that lists PC-98x1 memory and I/O ports by function and contains materials about different hardware aspects of the PC-98 platform (albeit in Japanese). 

The author of Neko Project 21/W has also compiled research materials (also in Japanese) he came across while working on the emulator, though looking at [the emulator source](https://drive.google.com/file/d/1_riYRB-iaX5buxaqaWSvY7hUNLCMowhz/view?usp=share_link) could shred more light. Here's a list of them, linked for convenience:

- [MELCO LGY-98 Analysis Materials](https://simk98.github.io/np21w/docs/lgy98.html)
- [Cirrus Logic CL-GD54xx Analysis Materials](https://simk98.github.io/np21w/docs/clgd54xx.html) (complementary to [UNDOCUMENTED 9801/9821 Vol. 2 - Window Accelerator Board](http://www.webtech.co.jp/company/doc/undocumented_mem/io_wab.txt))
- [Melco WSN-A2F/A4F Analysis Materials](https://simk98.github.io/np21w/docs/wsn.html) (also available [here](https://drive.google.com/file/d/1mfKgsAtqHZiRRm07hVsYBbct-P3kN90z/view?usp=sharing))
- [I-O Data GA-98NBI/C, II, IV Analysis Materials](https://simk98.github.io/np21w/docs/ga98nb.html)

Hope these could be helpful.

**Comment by cartifanwlr** [comment]
> The DEC Rainbow would also require Z80 emulation in addition to the 8088. 

That's true. We do emulate 8080 (which is the basis of Z80) as part of V20/30, so maybe it's possible, but dual CPU setups would be quite tricky. (Z-100 also suffers from this)

**Comment by Torinde** [comment]
> We do emulate 8080

"i8080 emulation isn't supported in 86Box yet." per @Cacodemon345 at https://github.com/joncampbell123/dosbox-x/issues/3853#issuecomment-1328977864

**Comment by planetbrian** [comment]
How about the Sun386i? It has Sun-3 features and runs a unique version of SunOS (4).

**Comment by TC1995** [comment]
The Sun386i also requires a TIGA chip for the Roadracer framebuffer.

**Comment by Torinde** [comment]
Tandy 2000, not PC compatible #2710

**Comment by iamgreaser** [comment]
Japanese might be difficult but it's not impossible. 日本語は難しくてもできないことじゃないですね。

1, 2: Not a problem. If you want to emulate a PC-98, you accept that it's mostly going to be in Japanese.

3: There is documentation available, at least for the older PC-98 machines. Furthermore, many of the chips used in the PC-98 are the same as used in the IBM PC architecture, although the I/O map is completely different and the memory map is... different enough to be incompatible, but seemingly inspired by IBM. The biggest difference is that the PC-98 started out with a 16-bit data bus, which is why 8-bit I/O peripherals tend to use the same least-significant bit for a given chip.

4: A way to map keys will always be necessary here. I have an OADG109A keyboard (I do not recommend getting one if you live outside of Japan unless you need to test things, it's a terrible idea), but even though that mostly covers the typing keys of the PC-98 keyboard, a lot of the other keys are quite different. (15 function keys, anyone?)

A big challenge may be correctly mapping a PC-98 BIOS image to a specific machine one wishes to emulate. As for working out what each machine's specs are, "PC-9800Series Technical Data Book - Hardware volume - revised edition" (PC-9800シリーズ テクニカルデータブック HARDWARE編 改訂版) was released in 1993 and seems to cover everything pertaining to that up to a point, including 2 Pentium models. The tables start on page 6, although unfortunately the Internet Archive gets the 2up pages offset incorrectly so it can be a bit of a pain to read from there: https://archive.org/details/NEC_PC-9800TechnicalDataBookHARDWAREOCR_1993/page/5/mode/2up

With that said, is anyone currently working on PC-98 support? If you have any branches for this then please link them, I'd like to help out and dig through some documentation for you. I can be found on the 86Box IRC channel as GreaseMonkey.

**Comment by OBattler** [comment]
@TC1995 has some preliminary work on his end. Since you're on our Discord server, just ask him there.
